tar/format.go

```diff
@@ -0,0 +1,185 @@
+package tarfmt
+
+import (
+	"archive/tar"
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	importer "github.com/ipfs/go-ipfs/importer"
+	chunk "github.com/ipfs/go-ipfs/importer/chunk"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	uio "github.com/ipfs/go-ipfs/unixfs/io"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+)
+
+var blockSize = 512
+var zeroBlock = make([]byte, blockSize)
+
+func marshalHeader(h *tar.Header) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	w := tar.NewWriter(buf)
+	err := w.WriteHeader(h)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func ImportTar(r io.Reader, ds dag.DAGService) (*dag.Node, error) {
+	rall, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	r = bytes.NewReader(rall)
+
+	tr := tar.NewReader(r)
+
+	root := new(dag.Node)
+	root.Data = []byte("ipfs/tar")
+
+	for {
+		h, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+
+		header := new(dag.Node)
+
+		headerBytes, err := marshalHeader(h)
+		if err != nil {
+			return nil, err
+		}
+
+		header.Data = headerBytes
+
+		if h.Size > 0 {
+			spl := chunk.NewRabin(tr, uint64(chunk.DefaultBlockSize))
+			nd, err := importer.BuildDagFromReader(ds, spl, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			err = header.AddNodeLinkClean("data", nd)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		_, err = ds.Add(header)
+		if err != nil {
+			return nil, err
+		}
+
+		err = root.AddNodeLinkClean(h.Name, header)
```

Comment:
so the root is going to get massive (thousands of links). we dont have sharding yet, so maybe this should work like the path bubble-up